pkg/provider/infisical/api/api.go

```diff
@@ -42,34 +43,99 @@ type InfisicalApis interface {
 	RevokeAccessToken() error
 }
 
+const (
+	machineIdentityLoginViaUniversalAuth = "MachineIdentityLoginViaUniversalAuth"
+	getSecretsV3                         = "GetSecretsV3"
+	getSecretByKeyV3                     = "GetSecretByKeyV3"
+	revokeAccessToken                    = "RevokeAccessToken"
+)
+
 const UserAgentName = "k8-external-secrets-operator"
-const errJSONSecretUnmarshal = "unable to unmarshal secret: %w"
 
-func NewAPIClient(baseURL string) (*InfisicalClient, error) {
+var errJSONUnmarshal = errors.New("unable to unmarshal API response")
+var errNoAccessToken = errors.New("unexpected error: no access token available to revoke")
+var errAccessTokenAlreadyRetrieved = errors.New("unexpected error: access token was already retrieved")
+
+type InfisicalAPIError struct {
+	StatusCode int
+	Err        any
+	Message    any
+	Details    any
+}
+
+func (e *InfisicalAPIError) Error() string {
+	if e.Details != nil {
+		detailsJSON, _ := json.Marshal(e.Details)
+		return fmt.Sprintf("API error (%d): error=%v message=%v, details=%s", e.StatusCode, e.Err, e.Message, string(detailsJSON))
+	} else {
+		return fmt.Sprintf("API error (%d): error=%v message=%v", e.StatusCode, e.Err, e.Message)
+	}
+}
+
+// checkError checks for an error on the http response and generates an appropriate error if one is
+// found.
+func checkError(resp *http.Response) error {
+	if resp.StatusCode < 400 {
+		return nil
+	}
+
+	respBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("API error (%d) and failed to read response body: %w", resp.StatusCode, err)
+	}
+
+	// Attempt to unmarshal the response body into an InfisicalAPIErrorResponse.
+	var errRes InfisicalAPIErrorResponse
+	err = ReadAndUnmarshal(io.NopCloser(bytes.NewBuffer(respBody)), &errRes)
+	// Non-200 errors that cannot be unmarshaled must be handled, as errors could come from outside of
+	// Infisical.
+	if err != nil {
+		return fmt.Errorf("API error (%d), could not unmarshal error response: %w", resp.StatusCode, err)
+	} else if reflect.DeepEqual(errRes, InfisicalAPIErrorResponse{}) {
```

Comment:
This should be `errors.Is` instead not a DeepEqual check.

Reply:
What I was trying to do here was check if the unmarshal was a zero-value. The `InfisicalAPIErrorResponse` is just a struct for the response and does not implement `Error`.

Is there a better way to do this? Definitely agreed `reflect` is rather suspicious here 😅

Reply:
This case could be a bit too defensive but the situation I'm imagining is where systems between Infisical (e.g. an ingress, gateway, etc.) are throwing the error, and the result won't be very useful other than the status code.

Reply:
Create a typed error a sentinel error and return that in certain cases. And once that's done, you can do an errors.Is there. Look at other places, search for errors.Is for such things. :) I have limited time right now elaborate, sorry, I'll get back to this.

Reply:
No problem, I really appreciate the help and certainly no rush!

I believe I sorted out a reasonable alternative and pushed it up.